### PR TITLE
Change os.rename to shutil.move

### PR DESCRIPTION
**Pull request by psmit**
This method was failing if the tmp directory is on a different filesystem. shutil.move tests for this and copies/removes if rename is not possible.

### CODE REVIEW

**Comment by macisamuele** [comment]
@psmit thanks for contributing to the project.
One information that I'l like to surface is that the usage of  `os.rename` is intentional and made because we need to somehow ensure atomicity of the file movement.

Could you eventually post the traceback of the error that you would get?

**Comment by psmit** [comment]
Sure:
```
Google Java Formatter....................................................Failed
- hook id: pretty-format-java
- exit code: 1

[cwd=/work/velo-accounting-broker] Run command: ('java', '-version')
[return_code=0] | openjdk version "15.0.2" 2021-01-19
OpenJDK Runtime Environment (build 15.0.2+7)
OpenJDK 64-Bit Server VM (build 15.0.2+7, mixed mode)

Downloading https://github.com/google/google-java-format/releases/download/v1.10.0/google-java-format-1.10.0-all-deps.jar
Traceback (most recent call last):
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/lib/python3.9/site-packages/language_formatters_pre_commit_hooks/pretty_format_java.py", line 37, in __download_google_java_formatter_jar
    return download_url(url_to_download, "google-java-formatter{version}.jar".format(version=version))
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/lib/python3.9/site-packages/language_formatters_pre_commit_hooks/utils.py", line 75, in download_url
    os.rename(tmp_file_name, final_file)
OSError: [Errno 18] Invalid cross-device link: '/tmp/tmpkte_a1ut' -> '/work/.cache/pre-commit/google-java-formatter1.10.0.jar'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/bin/pretty-format-java", line 8, in <module>
    sys.exit(pretty_format_java())
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/lib/python3.9/site-packages/language_formatters_pre_commit_hooks/pre_conditions.py", line 55, in wrapper
    return f(*args, **kwargs)
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/lib/python3.9/site-packages/language_formatters_pre_commit_hooks/pretty_format_java.py", line 79, in pretty_format_java
    google_java_formatter_jar = __download_google_java_formatter_jar(
  File "/work/.cache/pre-commit/repomn5h781s/py_env-python3.9/lib/python3.9/site-packages/language_formatters_pre_commit_hooks/pretty_format_java.py", line 45, in __download_google_java_formatter_jar
    raise RuntimeError(
RuntimeError: Failed to download any of: https://github.com/google/google-java-format/releases/download/v1.10.0/google-java-format-1.10.0-all-deps.jar, https://github.com/google/google-java-format/releases/download/google-java-format-1.10.0/google-java-format-1.10.0-all-deps.jar. Probably the requested version, 1.10.0, is not valid or you have some network issue.
```
Why do you need this move to be atomic?

**Comment by kylebalston-intel** [comment]
I'm seeing the same issue. One other possible option: we could leave the `os.rename` and instead add a `dir=base_directory` to the `NamedTemporaryFile` 

```python
    print("Downloading {url}".format(url=url), file=sys.stderr)
    r = requests.get(url, stream=True)
    r.raise_for_status()
    with tempfile.NamedTemporaryFile(dir=base_directory, delete=False) as tmp_file:  # Not delete because we're renaming it
        tmp_file_name = tmp_file.name
        shutil.copyfileobj(r.raw, tmp_file)
        tmp_file.flush()
        os.fsync(tmp_file.fileno())

    os.rename(tmp_file_name, final_file)
```

Since the temporary file would be alongside the final file, it should always be on the same device.

Though if we're not downloading to `$TMPDIR`, then perhaps we'd want to wrap the entire block in a try-except, so that we have a chance to clean up the temporary file created if an exception is thrown.

**Comment by macisamuele** [comment]
I kinda like more the approach suggested by @kylebalston-intel . It does not alter the logic and keep the same safety of atomic swap of the file.

@psmit would you have time to try it out to check if it accommodate your setup?

**Comment by psmit** [comment]
Yes, this would work! Shall I change this PR with this code? Or @kylebalston-intel would you want to make a new PR with this change and I'll close this one?

**Comment by kylebalston-intel** [comment]
@psmit -- if you could take this on, that would be great 👍

**Comment by psmit** [comment]
Sure, I have now pushed the fix you suggested. Unfortunately, I can't run the pipelines myself, I have to wait for a maintainer to approve them.

**Comment by macisamuele** [comment]
thanks a lot for contributing. tests are running now and if green this PR will be merged, release will then happen during the weekend. Still using the pre-commit using the latest available sha should give you access to the change

about the fact that tests are not run I'll need to check the settings, this is the first time I see this happening

**Review by macisamuele — APPROVED** [review]
change looks good to me. let's see the result of tests and if they are green then we can merge this.

thanks a lot for your contribution

**Comment by psmit** [comment]
Unfortunately, it seems that one test failed. However, it doesn't seem that this is related to this code....

**Comment by macisamuele** [comment]
@psmit I'll be looking and fixing it as needed. Thanks again for supporting

**Comment by macisamuele** [comment]
Sorry for the long latency on this.
Going to merge the PR as is as the test pre-commit failure was caused by some issues on the master branch addressed by #71.
Thanks a lot for the contributions

**Pull request merged**
